src/sage/combinat/permutation.py

```diff
@@ -3589,17 +3588,9 @@ def descent_polynomial(self):
             (the descent monomial in their (7.23) is different).
         """
         p = self
-        z = []
-        P = PolynomialRing(ZZ, len(p), 'z')
+        P = PolynomialRing(ZZ, len(p) + 1, 'z')
         z = P.gens()
-        result = 1
-        pol = 1
-        for i in range(len(p)-1):
-            pol *= z[p[i]-1]
-            if p[i] > p[i+1]:
-                result *= pol
-
-        return result
+        return P.prod(z[p[i]] for i in range(len(p) - 1) if p[i] > p[i + 1])
 
```

Comment:
The "TODO" needs to be updated. I agree that "descent polynomial" is not defined in GS1984. So that part can remain.